com.unity.netcode.gameobjects/Runtime/Serialization/BufferSerializer.cs

```diff
@@ -0,0 +1,313 @@
+using System;
+using Unity.Multiplayer.Netcode;
+using UnityEngine;
+
+namespace Unity.Netcode.Serialization
+{
+    /// <summary>
+    /// Two-way serializer wrapping FastBufferReader or FastBufferWriter.
+    /// 
+    /// Implemented as a ref struct for two reasons:
+    /// 1. The BufferSerializer cannot outlive the FBR/FBW it wraps or using it will cause a crash
+    /// 2. The BufferSerializer must always be passed by reference and can't be copied
+    ///
+    /// Ref structs help enforce both of those rules: they can't out live the stack context in which they were
+    /// created, and they're always passed by reference no matter what.
+    ///
+    /// BufferSerializer doesn't wrapp FastBufferReader or FastBufferWriter directly because it can't.
+    /// ref structs can't implement interfaces, and in order to be able to have two different implementations with
+    /// the same interface (which allows us to avoid an "if(IsReader)" on every call), the thing directly wrapping
```

Comment:
I'm trying to understand how we would end up using the BufferSerializer and IBufferSerializerImplementation. From what I can tell currently they are unused.

So if we were to replace `NetworkSerializer` with `BufferSerializer` we would have to pass the interface to the `NetworkSerialize` function of `INetworkSerializable`. This will incur boxing which is much worse than a `if(IsReader)` check which most of the time will be optimized away by branch prediction anyways.

Reply:
Yes, everything in this PR is currently unused. 

The plan is for INetworkSerializable's interface to change to be like this: 

```c#
interface INetworkSerializable
{
    void Serialize<T>(BufferSerializer<T>) where T: IBufferSerializerImplemetation;
}
```
BufferSerializer is a ref struct - c# language rules say it can NEVER be boxed. The generic function signature not only prevents boxing, but is actually the only way you can ever receive this type. And since `m_Implementation` is of a generic type rather than of IBufferSerializerImplementation, it also won't be boxed.

Reply:
The above code you sent won't work right? what is needed would be:

```cs
interface INetworkSerializable<T>
{
    void Serialize<T>(BufferSerializer<T>) where T: IBufferSerializerImplemetation;
}
```

And then as a user I would need to implement this generic interface for all instances so essentially:

```cs
public struct Foo: INetworkSerializable<BufferSerializerReader>, INetworkSerializable<BufferSerializerWriter>
{
    public int Bar;

    public void Serialize<BufferSerializerReader>(BufferSerializer<BufferSerializerReader> serializer)
    {
       serializer.Serialize(ref Bar);
    }

    public void Serialize<BufferSerializerWriter>(BufferSerializer<BufferSerializerWriter> serializer)
    {
       serializer.Serialize(ref Bar);
    }
}
```

And sure this works but it's ugly and defeats the point of having a serializer interface instead of a reader/writer interface.

Reply:
+1, we should continue to have a ReaderWrite interface if at all possible because it makes serialization code symmetrical by default, which is a huge win for users writing custom serializers

Reply:
No, the code I sent does work:

```c#
        interface IBufferSerializable
        {
            void Serialize<T>(BufferSerializer<T> serializer) where T : IBufferSerializerImplementation;
        }

        class Foo : IBufferSerializable
        {
            public void Serialize<T>(BufferSerializer<T> serializer) where T : IBufferSerializerImplementation
            {
                bool b = true;
                serializer.SerializeValue(ref b);
            }
        }

        static void SerializeIBufferSerializable(IBufferSerializable b)
        {
            FastBufferWriter writer = new FastBufferWriter(100, Allocator.Temp);
            BufferSerializer<BufferSerializerWriter> serializer =
                new BufferSerializer<BufferSerializerWriter>(new BufferSerializerWriter(ref writer));
            b.Serialize(serializer);
        }
```

It's still a read/write interface. It's still symmetrical. I just wrote that code in rider and it compiled fine, and no boxing.

Reply:
I promise it works. I tested it before I submitted the PR. I didn't keep that particular test code because I wasn't ready to change the definition of INetworkSerializable yet, but it does work.

Reply:
Ah I see, you are totally right, somehow missed that you can just keep the function as generic in the interface. Ok I'm completely fine with this.